helix-rest/src/main/java/org/apache/helix/rest/metadatastore/ZkMetadataStoreDirectory.java

```diff
@@ -111,6 +111,16 @@ public ZkMetadataStoreDirectory(Map<String, String> routingZkAddressMap)
     return allShardingKeys;
   }
 
+  @Override
+  public Map<String, List<String>> getShardingKeysByRealm(String namespace) {
+    Map<String, List<String>> shardingKeysByRealmMap = _realmToShardingKeysMap.get(namespace);
+    if (shardingKeysByRealmMap == null) {
+      throw new NoSuchElementException("Namespace " + namespace + " does not exist!");
+    }
+
+    return shardingKeysByRealmMap;
```

Comment:
```
if (!_realmToShardingKeysMap.contains(namespace)) {
  throw Exception(~~~);
}
return  _realmToShardingKeys.get(namespace);
```

seems like a more concise way of expressing this.

Reply:
I realized that we already consider multi-thread issue (reading and refreshing) for this map because it is a concurrent map. Your style `contains()` and `get()` would potentially have race condition. So only one `get()` makes advantage of a concurrent map to avoid race condition.

Reply:
I actually don’t see your suggested version is simpler: the key is hashed two times (maybe JVM could cache hashcode) which causes more cpu time. It is more natural to just compute one time of key hash. Right, in our case here namespace could not change, but it is just a special case here. What if namespace change could be supported in the future? I would prefer to do it in a general way: one time hash and cache the result and generally avoid race condition. So decouple it from namespace change.

Reply:
Sure - in this case either way works. Marking it as resolved 👍